File: crates/clang-importer-sys/cxx/lib/cxx/swift/ASTContext.cxx
#include "cxx/swift/ASTContext.hxx"

namespace cxx {
namespace swift {
namespace ASTContext {

std::unique_ptr<::swift::ASTContext>
get(rust::swift::LangOptions &langOpts,
    rust::swift::TypeCheckerOptions &typeckOpts,
    rust::swift::SILOptions &silOpts,
    rust::swift::SearchPathOptions &SearchPathOpts,
    rust::swift::ClangImporterOptions &ClangImporterOpts,
    rust::swift::symbolgraphgen::SymbolGraphOptions &SymbolGraphOpts,
    rust::swift::SourceManager &SourceMgr,
    rust::swift::DiagnosticEngine &Diags) {
  auto raw = ::swift::ASTContext::get(
      *langOpts.ptr, *typeckOpts.ptr, *silOpts.ptr, *SearchPathOpts.ptr,
      *ClangImporterOpts.ptr, *SymbolGraphOpts.ptr, *SourceMgr.ptr, *Diags.ptr);
  return std::unique_ptr<::swift::ASTContext>(raw);
}

std::unique_ptr<::swift::ASTContext> getWithCallback(
    rust::swift::LangOptions &langOpts,
    rust::swift::TypeCheckerOptions &typeckOpts,
    rust::swift::SILOptions &silOpts,
    rust::swift::SearchPathOptions &SearchPathOpts,
    rust::swift::ClangImporterOptions &ClangImporterOpts,
    rust::swift::symbolgraphgen::SymbolGraphOptions &SymbolGraphOpts,
    rust::swift::SourceManager &SourceMgr, rust::swift::DiagnosticEngine &Diags,
    rust::Fn<bool(rust::llvm::StringRef, bool)> PreModuleImportCallback) {
  std::function<bool(::llvm::StringRef, bool)> lambda = [=](::llvm::StringRef s,
                                                            bool b) {
    auto ptr = std::make_unique<::llvm::StringRef>(s);
    rust::llvm::StringRef t = {std::move(ptr)};
    return (*PreModuleImportCallback)(std::move(t), std::move(b));
  };
  auto raw = ::swift::ASTContext::get(
      *langOpts.ptr, *typeckOpts.ptr, *silOpts.ptr, *SearchPathOpts.ptr,
      *ClangImporterOpts.ptr, *SymbolGraphOpts.ptr, *SourceMgr.ptr, *Diags.ptr,
      lambda);
  return std::unique_ptr<::swift::ASTContext>(raw);
}

} // namespace ASTContext
} // namespace swift
} // namespace cxx

File: crates/clang-importer-sys/tests/swift/ast_context.rs
use clang_importer_sys::swift::{
    symbolgraphgen::SymbolGraphOptions,
    ASTContext,
    ClangImporterOptions,
    DiagnosticEngine,
    LangOptions,
    SILOptions,
    SearchPathOptions,
    SourceManager,
    TypeCheckerOptions,
};

#[test]
fn new() {
    #![allow(non_snake_case)]
    let mut lang_opts = LangOptions::new();
    let mut typeck_opts = TypeCheckerOptions::new();
    let mut sil_opts = SILOptions::new();
    let mut search_path_opts = SearchPathOptions::new();
    let mut clang_importer_opts = ClangImporterOptions::new();
    let mut symbol_graph_opts = SymbolGraphOptions::new();
    let mut source_mgr = SourceManager::new();
    let mut diags = DiagnosticEngine::from(&mut source_mgr);
    let _ = ASTContext::get(
        &mut lang_opts,
        &mut typeck_opts,
        &mut sil_opts,
        &mut search_path_opts,
        &mut clang_importer_opts,
        &mut symbol_graph_opts,
        &mut source_mgr,
        &mut diags,
    );
}
